### PR TITLE
Refactor cellery run command

### PR DESCRIPTION
**Pull request by MadhukaHarith92**
Refactor cellery run command

### CODE REVIEW

File: components/cli/ballerina/bal_executor.go [hunk 1]
```diff
@@ -20,19 +20,27 @@ package ballerina
...
 	"strings"
+	"sync"
 
 	"github.com/cellery-io/sdk/components/cli/pkg/constants"
+	"github.com/cellery-io/sdk/components/cli/pkg/image"
 	"github.com/cellery-io/sdk/components/cli/pkg/util"
 )
 
 const ballerina = "ballerina"
+const celleryEnvVar = "cellery_env_"
```
Comment (Mirage20) [inline components/cli/ballerina/bal_executor.go]: Is this a prefix?

File: components/cli/ballerina/bal_executor.go [hunk 2]
```diff
@@ -20,19 +20,27 @@ package ballerina
...
+const celleryImageDirEnvVar = "CELLERY_IMAGE_DIR"
 
 type BalExecutor interface {
 	Build(fileName string, iName []byte) error
+	Run(imageDir string, instanceName string, runningNode *DependencyTreeNode,
```
Comment (Mirage20) [inline components/cli/ballerina/bal_executor.go]: Does BalExecutor abstraction need to worry about cellery related configurations? like running node, shareDependencies flag?

File: components/cli/ballerina/bal_executor.go [hunk 3]
```diff
@@ -64,8 +72,55 @@ func (baleExecutor *LocalBalExecutor) Build(fileName string, iName []byte) error
 	return nil
 }
 
+// Run executes ballerina run on an executable bal file.
+func (balExecutor *LocalBalExecutor) Run(imageDir string, instanceName string, runningNode *DependencyTreeNode,
+	envVars []*EnvironmentVariable, tempRunFileName string, dependencyLinks map[string]*DependencyInfo,
+	startDependencies, shareDependencies bool) error {
+	cmd := &exec.Cmd{}
```
Comment (Mirage20) [inline components/cli/ballerina/bal_executor.go]: Do we need this initialization?

File: components/cli/ballerina/docker_bal_executor.go [hunk 4]
```diff
@@ -137,12 +137,117 @@ func (baleExecutor *DockerBalExecutor) Build(fileName string, iName []byte) erro
...
 }
 
+// Run executes ballerina run when ballerina is not installed.
+func (balExecutor *DockerBalExecutor) Run(imageDir string, instanceName string, runningNode *DependencyTreeNode,
+	envVars []*EnvironmentVariable, tempRunFileName string, dependencyLinks map[string]*DependencyInfo,
+	startDependencies, shareDependencies bool) error {
+
+	currentDir, err := os.Getwd()
+	if err != nil {
+		util.ExitWithErrorMessage("Error in determining working directory", err)
```
Comment (Mirage20) [inline components/cli/ballerina/docker_bal_executor.go]: Please return the error

File: components/cli/ballerina/docker_bal_executor.go [hunk 5]
```diff
@@ -137,12 +137,117 @@ func (baleExecutor *DockerBalExecutor) Build(fileName string, iName []byte) erro
...
+	}
+
+	//Retrieve the cellery cli docker instance status.
+	cmdDockerPs := exec.Command("docker", "ps", "--filter", "label=ballerina-runtime="+version.BuildVersion(),
+		"--filter", "label=currentDir="+currentDir, "--filter", "status=running", "--format", "{{.ID}}")
+
+	containerId, err := cmdDockerPs.Output()
+	if err != nil {
+		util.ExitWithErrorMessage("Docker Run Error", err)
```
Comment (Mirage20) [inline components/cli/ballerina/docker_bal_executor.go]: Same. Please check ExitWithErrorMessage in all places and return the error

File: components/cli/pkg/commands/run.go [hunk 6]
```diff
@@ -19,71 +19,58 @@
...
+	if err = cli.ExecuteTask("Extracting cell image", "Failed to extract cell image",
+		"", func() error {
+			imageDir, err = ExtractImage(parsedCellImage, true)
+			return err
+		}); err != nil {
+		return err
+	}
+	defer func() error {
+		if err = os.RemoveAll(imageDir); err != nil {
+			return fmt.Errorf("error occurred while extracting image, %v", err)
```
Comment (Mirage20) [inline components/cli/pkg/commands/run.go]: Is this error related to RemoveAll?

File: components/cli/ballerina/bal_executor.go (hunk 1 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Yes.

File: components/cli/ballerina/bal_executor.go (hunk 2 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Updated interface function method signature to be generic without cellery related configurations.

File: components/cli/ballerina/bal_executor.go (hunk 3 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Removed.

File: components/cli/ballerina/docker_bal_executor.go (hunk 4 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: Updated in all places.

File: components/cli/ballerina/docker_bal_executor.go (hunk 5 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: Updated in all places.

File: components/cli/pkg/commands/run.go (hunk 6 above)
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/run.go]: Yes.

File: components/cli/ballerina/docker_bal_executor.go [hunk 7]
```diff
@@ -137,12 +139,101 @@ func (baleExecutor *DockerBalExecutor) Build(fileName string, iName []byte) erro
...
+
+			_, errUserAdd := cmdUserAdd.Output()
+			if errUserAdd != nil {
+				return fmt.Errorf("error in adding Cellery execution user, %v", err)
+			}
+		}
+		exeUid = cliUser.Uid
+	}
+	re := regexp.MustCompile(`^.*cellery-cell-image`)
+	balFile = re.ReplaceAllString(balFile, "/home/cellery/.cellery/tmp/cellery-cell-image")
```
Comment (isurulucky) [inline components/cli/ballerina/docker_bal_executor.go]: move to a constant and refer from all hard coded places.

File: components/cli/pkg/commands/run.go (hunk 6 above)
Comment (Mirage20) [reply components/cli/pkg/commands/run.go]: But it does not say anything about the file removing

Comment (MadhukaHarith92) [reply components/cli/pkg/commands/run.go]: Updated the error message to "error occurred while removing image".

File: components/cli/ballerina/docker_bal_executor.go [hunk 8]
```diff
@@ -137,12 +137,117 @@ func (baleExecutor *DockerBalExecutor) Build(fileName string, iName []byte) erro
...
+	if string(containerId) == "" {
+		cmdDockerRun := exec.Command("docker", "run", "-d", "-l", "ballerina-runtime="+version.BuildVersion(),
+			"--mount", "type=bind,source="+util.UserHomeDir()+string(os.PathSeparator)+".ballerina,target=/home/cellery/.ballerina",
+			"--mount", "type=bind,source="+util.UserHomeDir()+string(os.PathSeparator)+".cellery,target=/home/cellery/.cellery",
+			"--mount", "type=bind,source="+util.UserHomeDir()+string(os.PathSeparator)+".kube,target=/home/cellery/.kube",
+			"wso2cellery/ballerina-runtime:"+version.BuildVersion(), "sleep", "600",
+		)
+		containerId, err = cmdDockerRun.Output()
+		if err != nil {
+			util.ExitWithErrorMessage("Docker Run Error", err)
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: This is outdated.

File: components/cli/pkg/commands/run.go [hunk 9]
```diff
@@ -141,261 +130,148 @@ func RunRun(cellImageTag string, instanceName string, startDependencies bool, sh
...
-	tempPath, err := ioutil.TempDir(celleryHomeTmp, "cellery-cell-image")
-	if err != nil {
+	var tempPath string
+	if tempPath, err = ioutil.TempDir(celleryHomeTmp, "cellery-cell-image"); err != nil {
 		return "", err
 	}
-	err = util.Unzip(zipLocation, tempPath)
-	if err != nil {
+	if err = util.Unzip(zipLocation, tempPath); err != nil {
 		return "", nil
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/run.go]: Updated.

File: components/cli/ballerina/docker_bal_executor.go (hunk 7 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: Updated.
